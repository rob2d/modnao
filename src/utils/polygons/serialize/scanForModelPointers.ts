import TransferrableBuffer from '@/types/TransferrableBuffer';

/**
 * @returns an array of pointers for each available
 * model discovered
 */
export default function scanForModelPointers(
  buffer: TransferrableBuffer
): [
  modelPointers: { address: number; ramAddress: number }[],
  modelRamOffset: number
] {
  const wBuffer = Buffer.from(buffer);
  // get initial offset that will dictate how we can pull
  // textures/model data from files since typically they are loaded into RAM
  // before these ops in actual game-process on SH4
  const modelRamOffset = wBuffer.readUInt32LE(0x0000) & 0xffffff00;
  const modelTablePointer = wBuffer.readUInt32LE(0x0000) - modelRamOffset;
  const modelCount = wBuffer.readUInt32LE(0x0004);
  const modelPointers: { address: number; ramAddress: number }[] = [];

  for (let i = 0; i < modelCount; i++) {
    const pointer = modelTablePointer + 4 * i;
    try {
      const ramAddress = wBuffer.readUInt32LE(pointer);
      const address = ramAddress - modelRamOffset;

      modelPointers.push({ address, ramAddress });
    } catch (error) {
      console.error(`Error: error occurred scanning model at index ${i}`);
      console.error(error);
    }
  }

  return [modelPointers, modelRamOffset];
}
