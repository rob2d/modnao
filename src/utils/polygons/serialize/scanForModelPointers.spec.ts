import { promises as fs } from 'fs';
import path from 'path';

import scanForModelPointers from './scanForModelPointers';

// @TODO use model generation logic when that exists vs
// arbitrary stage test files with more methodical approach

describe('scanForModelPointers', () => {
  it('returns the correct list of model pointer addresses in a stage file', async () => {
    const polygonFile = await fs.readFile(
      path.join(process.cwd(), 'src/__mocks__/STGXXPOL.BIN')
    );

    const buffer = new SharedArrayBuffer(polygonFile.byteLength);
    new Uint8Array(buffer).set(new Uint8Array(polygonFile));
    const results = scanForModelPointers(buffer);
    expect(results).toMatchSnapshot('scan-model-for-pointer-addresses');
  });
});
