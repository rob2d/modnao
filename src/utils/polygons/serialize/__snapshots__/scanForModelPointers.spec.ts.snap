// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scanForModelPointers returns the correct list of model pointer addresses in a stage file: scan-model-for-pointer-addresses 1`] = `
[
  [
    {
      "address": 528,
      "ramAddress": 216662544,
    },
    {
      "address": 1856,
      "ramAddress": 216663872,
    },
    {
      "address": 2784,
      "ramAddress": 216664800,
    },
    {
      "address": 3712,
      "ramAddress": 216665728,
    },
    {
      "address": 9240,
      "ramAddress": 216671256,
    },
    {
      "address": 14248,
      "ramAddress": 216676264,
    },
    {
      "address": 19256,
      "ramAddress": 216681272,
    },
    {
      "address": 19288,
      "ramAddress": 216681304,
    },
    {
      "address": 19320,
      "ramAddress": 216681336,
    },
    {
      "address": 19696,
      "ramAddress": 216681712,
    },
    {
      "address": 24704,
      "ramAddress": 216686720,
    },
    {
      "address": 29712,
      "ramAddress": 216691728,
    },
    {
      "address": 34992,
      "ramAddress": 216697008,
    },
    {
      "address": 35712,
      "ramAddress": 216697728,
    },
    {
      "address": 35960,
      "ramAddress": 216697976,
    },
    {
      "address": 40280,
      "ramAddress": 216702296,
    },
    {
      "address": 45288,
      "ramAddress": 216707304,
    },
    {
      "address": 50296,
      "ramAddress": 216712312,
    },
    {
      "address": 50544,
      "ramAddress": 216712560,
    },
    {
      "address": 55552,
      "ramAddress": 216717568,
    },
    {
      "address": 56016,
      "ramAddress": 216718032,
    },
    {
      "address": 60096,
      "ramAddress": 216722112,
    },
    {
      "address": 60344,
      "ramAddress": 216722360,
    },
    {
      "address": 65352,
      "ramAddress": 216727368,
    },
    {
      "address": 70360,
      "ramAddress": 216732376,
    },
    {
      "address": 74680,
      "ramAddress": 216736696,
    },
    {
      "address": 75400,
      "ramAddress": 216737416,
    },
    {
      "address": 76080,
      "ramAddress": 216738096,
    },
    {
      "address": 76976,
      "ramAddress": 216738992,
    },
    {
      "address": 78088,
      "ramAddress": 216740104,
    },
    {
      "address": 78336,
      "ramAddress": 216740352,
    },
    {
      "address": 79664,
      "ramAddress": 216741680,
    },
    {
      "address": 79912,
      "ramAddress": 216741928,
    },
    {
      "address": 80160,
      "ramAddress": 216742176,
    },
    {
      "address": 80408,
      "ramAddress": 216742424,
    },
    {
      "address": 80656,
      "ramAddress": 216742672,
    },
    {
      "address": 80904,
      "ramAddress": 216742920,
    },
    {
      "address": 81152,
      "ramAddress": 216743168,
    },
    {
      "address": 81400,
      "ramAddress": 216743416,
    },
    {
      "address": 81784,
      "ramAddress": 216743800,
    },
    {
      "address": 82096,
      "ramAddress": 216744112,
    },
    {
      "address": 82480,
      "ramAddress": 216744496,
    },
    {
      "address": 82864,
      "ramAddress": 216744880,
    },
    {
      "address": 83248,
      "ramAddress": 216745264,
    },
    {
      "address": 83560,
      "ramAddress": 216745576,
    },
    {
      "address": 83808,
      "ramAddress": 216745824,
    },
    {
      "address": 84056,
      "ramAddress": 216746072,
    },
    {
      "address": 84440,
      "ramAddress": 216746456,
    },
    {
      "address": 85120,
      "ramAddress": 216747136,
    },
    {
      "address": 85368,
      "ramAddress": 216747384,
    },
    {
      "address": 85616,
      "ramAddress": 216747632,
    },
    {
      "address": 92808,
      "ramAddress": 216754824,
    },
    {
      "address": 93488,
      "ramAddress": 216755504,
    },
    {
      "address": 94384,
      "ramAddress": 216756400,
    },
    {
      "address": 95496,
      "ramAddress": 216757512,
    },
    {
      "address": 96824,
      "ramAddress": 216758840,
    },
    {
      "address": 98368,
      "ramAddress": 216760384,
    },
    {
      "address": 100128,
      "ramAddress": 216762144,
    },
    {
      "address": 102104,
      "ramAddress": 216764120,
    },
    {
      "address": 104296,
      "ramAddress": 216766312,
    },
    {
      "address": 104544,
      "ramAddress": 216766560,
    },
    {
      "address": 104792,
      "ramAddress": 216766808,
    },
    {
      "address": 104824,
      "ramAddress": 216766840,
    },
    {
      "address": 104856,
      "ramAddress": 216766872,
    },
    {
      "address": 104888,
      "ramAddress": 216766904,
    },
    {
      "address": 105136,
      "ramAddress": 216767152,
    },
    {
      "address": 105168,
      "ramAddress": 216767184,
    },
    {
      "address": 105200,
      "ramAddress": 216767216,
    },
    {
      "address": 105448,
      "ramAddress": 216767464,
    },
    {
      "address": 105696,
      "ramAddress": 216767712,
    },
    {
      "address": 105944,
      "ramAddress": 216767960,
    },
    {
      "address": 105976,
      "ramAddress": 216767992,
    },
    {
      "address": 108320,
      "ramAddress": 216770336,
    },
    {
      "address": 108352,
      "ramAddress": 216770368,
    },
    {
      "address": 108384,
      "ramAddress": 216770400,
    },
    {
      "address": 108416,
      "ramAddress": 216770432,
    },
    {
      "address": 108448,
      "ramAddress": 216770464,
    },
    {
      "address": 108696,
      "ramAddress": 216770712,
    },
    {
      "address": 108944,
      "ramAddress": 216770960,
    },
    {
      "address": 109256,
      "ramAddress": 216771272,
    },
    {
      "address": 109288,
      "ramAddress": 216771304,
    },
    {
      "address": 109320,
      "ramAddress": 216771336,
    },
    {
      "address": 109352,
      "ramAddress": 216771368,
    },
  ],
  216662016,
]
`;
